contents/blog/plugins-part-1-the-world-map.md

```diff
@@ -0,0 +1,106 @@
+
+This is perfect for a hackathon demo, but will it work in real life? What are the unknown unknowns we don't know about? [IS THIS WEB SCALE?](http://www.mongodb-is-web-scale.com/)
+
+It didn't take long for cracks to appear.
+
+Turns out it's damn near impossible to manage Python dependencies via multiple `requirements.txt` files.
+
+There's no way to tell which dependencies your plugin's dependencies will install, without first installing them. There is [nothing like `pip --dry-run`](https://github.com/pypa/pip/issues/53). You literally need to install the `pip` package, and run an `__init__.py` script inside the package to get its dependency tree.
+
+What's more, all dependencies get installed together, making it just a matter of time before a plugin overrides an app dependency (e.g. `django`), and brings everything crashing down. Not cool.
+
+We had no choice but to scrap this work.
+
+Perhaps we should have gone with JavaScript from day one?
+
+### Other attempts
+
+As you can [read from the Plugins Epic](https://github.com/PostHog/posthog/issues/1896#issuecomment-717855986), I tried a few other options next:
+
+- I developed and then completely removed a JS-via-[py-mini-racer](https://github.com/sqreen/PyMiniRacer) implementation. Mini racer is great for simple functions, but as it's raw v8, not node, it lacks support for 1) async/promises, 2) importing modules like `fetch`, 3) very limited standard library. No bueno.
+- I developed and then completely removed a JS gRPC implementation. It worked well - I was able to call JS code from Python and get the response, but this approach leaves us with a lot of manual work: 1) how many workers should respond to the gRPC calls? 2) how to make sure no events are lost if the node gRPC server is down?
+
+Finally, after a bit of soul-searching, I came up with an ingenious plan.
+
+We use [Celery](https://docs.celeryproject.org/) to process events in the background. When an event hits `/capture`, our API parses the request, and queues the event onto a job queue.
+
+What if we built a new server with the [NodeJS port of Celery](https://celery-node.js.org/), and just plugged that in as another step in the existing pipeline? 
+
+This would solve all pending issues: we wouldn't have to worry about Python dependencies (it's JavaScript all the way down), there would be no process management for a gRPC link, and we could eventually rewrite the entire ingestion pipeline in JavaScript to get a speed boost over Python.
+
+What's not to like?
+
+Well, if it was only that simple...
+
+## Read next 
+
+This article is the first in a series about plugins in PostHog.
+
+- Part 1. The World Map.
+- Part 2. How to build a scalable arbitrary code execution machine.
+- Part 3. Rebuilding action-mapping to a real-time pipeline
+- Part 4. How plugins lower the barrier to entry for contributors
+- Part 5. ???
+- Part 6. Profit!
```

Comment:
The last two are funny, but now I'm not sure if this series is a joke or for real.